File: LinkedList/SinglyLinkedList/Medium/Add1toList.java
package Medium;

import x.*;

public class Add1toList {
    public static void main(String[] args) {
        LinkedList l1 = new LinkedList();
        l1.append(9);
        l1.append(9);
        l1.append(9);

        // reverse
        reversal(l1);
        // add 1 to 999
        addOne(l1);
        // reverse
        reversal(l1);
        l1.print();
    }

    static void addOne(LinkedList l) {
        Node cur = l.head, prev = null;
        int carry = 1;
        // 1->2->3->4
        // add carry to all values
        while (cur != null) {
            int sum = carry + cur.val;
            cur.val = sum % 10;
            carry = sum / 10;

            prev = cur;
            cur = cur.next;
        }

        if (carry == 1) {
            prev.next = new Node(1);
        }
    }

    static void reversal(LinkedList list) {
        // using 3 references
        // input: head->1->2->3->4->null
        // null<-1<-2<-3<-4<-head
        // output: null<-1<-2<-3<-4<-head
        Node current = list.head, prev = null, next = null;
        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        list.head = prev;
    }

}

File: LinkedList/SinglyLinkedList/Medium/Add1ToList.js
// work in progress
import { LinkedList, Node } from "../x/LinkedList.js";

const list = new LinkedList([ 9, 9, 9 ]);
list.print();

// 9->2->1
const add1ToList = (list) => {
	let cur = list.head;
	let node = new Node(1);
	// 9 -> 2 -> 1 -> null
	// 1
	let carry = 1;
	while (cur != null) {
		let sum = cur.val + carry;
		cur.val = sum % 10;
		carry = parseInt(sum / 10);

		cur = cur.next;
	}

	if (carry) {
		// append new node with 1 to the end of the list
		let node = new Node(1);
		let cur = list.head;
		while (cur.next != null) {
			cur = cur.next;
		}
		cur.next = node;
	}
};

add1ToList(list);
list.print();
